lib/puppet/util/windows/registry.rb

```diff
@@ -234,7 +234,7 @@ def read(key, name_ptr, *rtype)
         begin
           case type
             when Win32::Registry::REG_SZ, Win32::Registry::REG_EXPAND_SZ
-              result = [ type, sanitize(data_ptr.read_wide_string(string_length)) ]
+              result = [ type, data_ptr.read_arbitrary_wide_string_up_to(string_length) ]
```

Comment:
I have concerns about the `read_arbitrary_wide_string_up_to` method, see my comment in https://github.com/puppetlabs/puppet/pull/7943#issuecomment-579541839. I also don't like that `read_arbitrary_wide_string_up_to` uses [`send` to read either one or two bytes](https://github.com/puppetlabs/puppet/blob/9b5bd2ad8b180efa4c143798a1a2b362928f1f1c/lib/puppet/util/windows/api_types.rb#L80). I'm also not sure if it properly double-null terminates the string if it's [longer than the max](https://github.com/puppetlabs/puppet/blob/9b5bd2ad8b180efa4c143798a1a2b362928f1f1c/lib/puppet/util/windows/api_types.rb#L84)? So I was thinking of adding an optional argument to `read_wide_string`, something like `read_wide_string(length, strip = false)`. If `strip = true`, then have the `read_wide_string` method scan the wide string for a wide terminator, and if one is found, truncate/substring before [transcoding to UTF-8](https://github.com/puppetlabs/puppet/blob/9b5bd2ad8b180efa4c143798a1a2b362928f1f1c/lib/puppet/util/windows/api_types.rb#L59).

Reply:
My concerns may also be unfounded, so if it's doing the right thing, then go ahead and use it.

Reply:
I did some similar work with @oanatmaria for facter-ng, it could worth checking it: https://github.com/puppetlabs/facter-ng/blob/master/lib/resolvers/windows/ffi/ffi.rb#L36

Reply:
Seems like we could just `split` the string and return the first part:

```
irb(main):003:0> "foo\u0000bar".encode('UTF-16LE').split("\u0000".encode('UTF-16LE')).first
=> "foo"
irb(main):004:0> "foo".encode('UTF-16LE').split("\u0000".encode('UTF-16LE')).first
=> "foo"
```

I think the extra complexity with trying to handle either a single or double NULL in `read_arbitrary_wide_string_up_to` is not really needed.

Reply:
Seems like if we want to split, first we need to clear the invalid characters
```ruby
>> str #=> "\u3631\u302E\u312E\u3330\u3735\u322E\u3030\u3230\u0000\u0000\u8FE2\uEBB3\u8AB2\uBFE7\u00BA\u94E6\uEB90\uB4A5\uBFE7\u00BA\x90\xD8\uBDE6\u5995"
>> x = str.encode('UTF-16LE').split("\u0000".encode('UTF-16LE')).first
Traceback (most recent call last):
        9: from /Users/gheorghe.popescu/.rvm/gems/ruby-2.6.3/bin/ruby_executable_hooks:24:in `<main>'
        8: from /Users/gheorghe.popescu/.rvm/gems/ruby-2.6.3/bin/ruby_executable_hooks:24:in `eval'
        7: from /Users/gheorghe.popescu/.rvm/gems/ruby-2.6.3/bin/irb:23:in `<main>'
        6: from /Users/gheorghe.popescu/.rvm/gems/ruby-2.6.3/bin/irb:23:in `load'
        5: from /Users/gheorghe.popescu/.rvm/gems/ruby-2.6.3/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        4: from /Users/gheorghe.popescu/.rvm/gems/ruby-2.6.3/gems/fancy_irb-1.2.1/lib/fancy_irb/irb_ext.rb:19:in `signal_status'
        3: from /Users/gheorghe.popescu/.rvm/gems/ruby-2.6.3/gems/fancy_irb-1.2.1/lib/fancy_irb/irb_ext.rb:31:in `evaluate'
        2: from (irb):25
        1: from (irb):25:in `split'
ArgumentError (invalid byte sequence in UTF-16LE)
>> x = str.encode('UTF-16LE', invalid: :replace, replace: "".encode("UTF-16LE")).split("\u0000".encode('UTF-16LE')).first
>> x #=> "\u3631\u302E\u312E\u3330\u3735\u322E\u3030\u3230"
>> x.encode("UTF-8", x.encoding, {}) #=> "㘱〮ㄮ㌰㜵㈮〰㈰"
>>
```

Reply:
How about replacing the invalid and undef characters when doing the transcoding?
```
 result = [ type, sanitize(data_ptr.read_wide_string(string_length, Encoding::UTF_8, :invalid => :replace, :undef => :replace)) ]
```

Reply:
Scanning to the first "\u0000" and keep the substring before
```[1] pry(Puppet::Provider::Package::Windows::Package)> data_ptr.read_wide_string(string_length,  Encoding::UTF_8, true)
=> "16.0.10357.20002"
[2] pry(Puppet::Provider::Package::Windows::Package)>
```


```ruby
def read_wide_string(char_length, dst_encoding = Encoding::UTF_8, strip = false, encode_options = {})
      # char_length is number of wide chars (typically excluding NULLs), *not* bytes
      str = get_bytes(0, char_length * 2).force_encoding('UTF-16LE')
     
      if strip == true
        i = str.index("\u0000".encode('UTF-16LE'))
        if i.nil?
          str = str
        else
          str = i == 0 ? "" : str[0..(i - 1)]
        end
      end

      str.encode(dst_encoding, str.encoding, encode_options)
    rescue Exception => e
      Puppet.debug "Unable to convert value #{str.nil? ? 'nil' : str.dump} to encoding #{dst_encoding} due to #{e.inspect}"
      raise
    end
```

Reply:
👍 to adding a `strip` parameter, but I'd drop the `encode_options` until if/when we need that.

I think you could reduce some of the difference cases (nil, 0, >0) to something like:

```ruby
if strip
  i = str.index(...)
  str = str[0, i] unless i.nil?
end
str.encode(dst_encoding)
```
Also I think we'd want to rescue `Encoding::EncodingError` instead of all `Exception`s.